Logging

JimMadge opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

- Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
- If you're not yet ready to merge, please mark this pull request as a **draft** and add `'[WIP]'` to the title
- Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

<!--
List any other PRs that must be merged before this one
-->

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

- Logging changes
  - Simplify handler/formatter construction
  - Remove log level, file, line number from console handler
- Move unrelated methods from logging class
- Remove top level try/except
- More Pythonic use of logging module (makes configuration neater, particularly in unit tests)


The approach here is

- ConsoleHandler: Logging to console, information for the user, INFO and above
- PlainFileHandler: Logging to a file with rich markup removed, information for auditing and debugging, all levels

with logging levels

- rich.print: When output is only of interest to the user. Used for commands which send their output to stdout.
- logger.info: Used for notifications of actions taken and completed successfully. stdout which is also useful to log.
- logger.warning: Used for unexpected events or potential problems that won't cause the program to stop. User should take notice even if the program continues.
- logger.error: Used for unexpected events or problems that may cause the program to stop.
- logger.fatal: Used for problems that will stop the program, immediately before an exit call.

### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #1891 
Closes #1657

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

JimMadge committed:
WIP: Simplify handlers

JimMadge committed:
Restore markup

JimMadge committed:
Remove output file argument

JimMadge committed:
Add verbose option

JimMadge committed:
Add show level option

JimMadge committed:
Move prompts

jemrobinson commented:
Something to consider (no problem if you decide not to) - should we log all runs to `$appdirs / $date-$user-something.log` and potentially also upload these to Azure for future reference?

JimMadge committed:
Move tabulate

JimMadge committed:
WIP: log errors on exceptions

JimMadge committed:
Fix circular dependency

JimMadge committed:
Sort imports

JimMadge committed:
Correct type hint

JimMadge committed:
Adjust log levels and printing

JimMadge committed:
Add log directory

JimMadge committed:
Use tz aware datetime

JimMadge committed:
Fix help tests

JimMadge committed:
Add version test

JimMadge committed:
Fix diff rendering

JimMadge committed:
Change log level of context switch

JimMadge committed:
Fix mocks

JimMadge committed:
Sort imports

JimMadge committed:
Add fixture for tmp logging dir

JimMadge committed:
Add test for stripping rich formatting

JimMadge committed:
Restructure logger, don't use singleton

The singleton structure was causing problems in unit tests.
This commit uses a more Pythonic style of logging (I wonder if the
reason Python logging works as it does it because they tried something
like we did an also found it introduced unexpected problems).

JimMadge committed:
Fix tests

JimMadge committed:
Add get_logger test

JimMadge committed:
Fix mypy errors

JimMadge committed:
Add set_console_level test

JimMadge committed:
Use package logger instead of root

JimMadge committed:
Add show_console_level test

JimMadge committed:
Add stdout tests

JimMadge committed:
Run lint:fmt

jemrobinson commented on data_safe_haven/commands/cli.py:
```diff
@@ -96,9 +97,4 @@ def callback(
 
 def main() -> None:
     """Run the application and log any exceptions"""
```
... and don't log exceptions? :)

jemrobinson commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -1,5 +1,12 @@
+from data_safe_haven.logging import get_logger
+
+
 class DataSafeHavenError(Exception):
-    pass
+    def __init__(self, message: str | bytes):
+        super().__init__(message)
+
+        logger = get_logger()
+        logger.error(message)
```
What does this do?

jemrobinson commented on data_safe_haven/logging/logger.py:
```diff
@@ -0,0 +1,93 @@
+"""Standalone logging class implemented as a singleton"""
```
This doesn't describe this file any more.

jemrobinson commented on data_safe_haven/utility/console.py:
```diff
@@ -0,0 +1,25 @@
+from rich import print as rprint
```
As this is still to do with printing information to screen, I think it could/should live in the logging module. Happy if you think that means we need to change the module name to something broader.

JimMadge replied:
I could see a module for "console stuff" like formatting text and prompts.

I think that is quite different from logging though and we should keep them separate.

I will leave it as is for this PR.

jemrobinson commented on data_safe_haven/utility/prompts.py:
```diff
@@ -0,0 +1,14 @@
+from rich.prompt import Confirm
```
As above, I'd like to see this live with the other logging commands.

jemrobinson commented on data_safe_haven/commands/cli.py:
```diff
@@ -1,13 +1,12 @@
 from typing import Annotated, Optional
 
 import typer
+from rich import print as rprint
```
I think we should replace this with a `console.print` function that passes its args through to `rich.print`. That would simplify our include structure and unify the things that write straight to the console.

JimMadge replied:
Do you mean [`rich.console.Console.print`](https://rich.readthedocs.io/en/stable/reference/console.html#rich.console.Console.print) or making something like `data_safe_haven.console.print`?

That could work with the other two comments here, introducing a `data_safe_haven.{console,output,terminal,text}` module to hold these things.

jemrobinson reviewed (approved):
Overall, LGTM. Can you post some examples of what the output looks like before and after for places where it's changed?

JimMadge committed:
Improve docstrings and comments

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge commented:
Merged through docs link check failing as that wasn't modified here. Does look like a genuine failure though.

JimMadge commented:
Screenshots,

- New with `--verbose --show-level`
- New (default)
- Old

<img width="749" alt="Screenshot 2024-06-20 at 11 15 26" src="https://github.com/alan-turing-institute/data-safe-haven/assets/23616154/6adee850-439b-481c-bd44-661ecc7485a3">

JimMadge commented:
And the log file,

```
2024-06-20 11:13:47,048 - DEBUG - Reading project settings from '/Users/jmadge/Library/Application Support/data_safe_haven/contexts.yaml'.
2024-06-20 11:13:47,067 - INFO - Switched context to 'elements'.
2024-06-20 11:13:47,073 - DEBUG - Saved context settings to '/Users/jmadge/Library/Application Support/data_safe_haven/contexts.yaml'.
2024-06-20 11:14:06,576 - DEBUG - Reading project settings from '/Users/jmadge/Library/Application Support/data_safe_haven/contexts.yaml'.
2024-06-20 11:14:06,586 - INFO - Switched context to 'elements'.
2024-06-20 11:14:06,595 - DEBUG - Saved context settings to '/Users/jmadge/Library/Application Support/data_safe_haven/contexts.yaml'.
```

jemrobinson commented:
Looks great. I like the idea of keeping timestamps in the log file but not in the console output (by default). Do you think there's any benefit in adding an option to include timestamps in the console output?

JimMadge commented:
> Looks great. I like the idea of keeping timestamps in the log file but not in the console output (by default). Do you think there's any benefit in adding an option to include timestamps in the console output?

That did occur to me when I posted the pictures. Could definitely do that too, should be an easy addition.
